src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+        const is_me = peer_contact_info.pubkey.equals(&my_pubkey);
+        if (is_me) {
+            result.is_me_count += 1;
+            continue;
+        }
+
+        const matching_shred_version = my_shred_version == peer_contact_info.shred_version or my_shred_version == 0;
+        if (!matching_shred_version) {
+            result.invalid_shred_version += 1;
+            continue;
+        }
+        _ = peer_contact_info.getSocket(socket_tag.RPC) orelse {
+            result.no_rpc_count += 1;
+            continue;
+        };
+        const gossip_data = table.get(.{ .SnapshotHashes = peer_contact_info.pubkey }) orelse {
+            result.no_snapshot_hashes_count += 1;
+            continue;
+        };
+        const snapshot_hashes = gossip_data.value.data.SnapshotHashes;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (snapshot_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // dont try to download from a slow peer
+        for (blacklist) |black_list_peers| {
+            if (black_list_peers.equals(&peer_contact_info.pubkey)) {
+                result.is_blacklist += 1;
+                continue :search_loop;
+            }
+        }
+
+        valid_peers.appendAssumeCapacity(.{
+            // NOTE: maybe we need to deep clone here due to arraylist sockets?
+            .contact_info = peer_contact_info.*,
+            .full_snapshot = snapshot_hashes.full,
+            .inc_snapshot = max_inc_hash,
+        });
+    }
+    result.is_valid = valid_peers.items.len;
+
+    return result;
+}
+
+/// uses swap remove to remove peers that do not have a hash which is trusted
+pub fn removeUnTrustedSnapshotPeers(
+    allocator: std.mem.Allocator,
+    table: *const GossipTable,
+    trusted_validators: std.ArrayList(Pubkey),
+    available_snapshot_peers: *std.ArrayList(PeerSnapshotHash),
+) !void {
+    var trusted_snapshot_hashes = std.AutoHashMap(
+        SlotAndHash, // full snapshot hash
+        std.AutoHashMap(SlotAndHash, void), // set of incremental snapshots
+    ).init(allocator);
+    defer trusted_snapshot_hashes.deinit();
+
+    // get the hashes of trusted validators
+    var trusted_count: usize = 0;
+    for (trusted_validators.items) |trusted_validator| {
+        const gossip_data = table.get(.{ .SnapshotHashes = trusted_validator }) orelse continue;
+        const trusted_hashes = gossip_data.value.data.SnapshotHashes;
+        trusted_count += 1;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (trusted_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // track the full and incremental
+        var r = try trusted_snapshot_hashes.getOrPut(trusted_hashes.full);
```

Comment:
This inserts an undefined value into the hashmap. If there are no incremental snapshots, it remains undefined.